service/src/test/java/bio/terra/workspace/pact/TpsApiTest.java

```diff
@@ -0,0 +1,555 @@
+            "/api/policy/v1alpha1/pao/%s".formatted(existingPolicyId))
+        .willRespondWith()
+        .status(HttpStatus.OK.value())
+        .body(updateResponseShape)
+        .toPact();
+  }
+
+  @Test
+  @PactTestFor(pactMethod = "updatePaoPreexistingNoConflicts")
+  public void updatingAnExistingPaoWithNoConflicts() throws Exception {
+    var remove =
+        new TpsPolicyInputs()
+            .inputs(List.of(new TpsPolicyInput().name("test_name").namespace("test_namespace")));
+    var add =
+        new TpsPolicyInputs()
+            .inputs(List.of(new TpsPolicyInput().name("new_name").namespace("new_namespace")));
+    var result = dispatch.updatePao(existingPolicyId, add, remove, TpsUpdateMode.FAIL_ON_CONFLICT);
+    assertNotNull(result);
+  }
+
+  @Pact(consumer = "wsm", provider = "tps")
+  public RequestResponsePact listValidRegions(PactDslWithProvider builder) {
+    return builder
+        .given(existingPolicyState)
+        .uponReceiving("A request to list the valid regions for a policy using the id")
+        .method("GET")
+        .pathFromProviderState(
+            "/api/policy/v1alpha1/region/%s/list-valid".formatted(existingPolicyProviderStateValue),
+            "/api/policy/v1alpha1/region/%s/list-valid".formatted(existingPolicyId))
+        .matchQuery("platform", cloudPlatformTpsRegex)
+        .willRespondWith()
+        .status(HttpStatus.OK.value())
+        .body(new PactDslJsonArray().stringType())
+        .toPact();
+  }
+
+  @Test
+  @PactTestFor(pactMethod = "listValidRegions")
+  public void listingValidRegionsOnAWorkspace() throws Exception {
+    var result = dispatch.listValidRegions(existingPolicyId, CloudPlatform.AZURE);
+    assertNotNull(result);
+  }
+
+  @Pact(consumer = "wsm", provider = "tps")
+  public RequestResponsePact listValidByPolicyInput(PactDslWithProvider builder) {
+    return builder
+        .given(existingPolicyState)
+        .uponReceiving("A request to list the valid regions for a policy using policy input")
+        .method("POST")
+        .body(tpsPolicyInputsObjectShape)
+        .path("/api/policy/v1alpha1/location/list-valid")
+        .matchQuery("platform", cloudPlatformTpsRegex)
+        .headers(contentTypeJsonHeader)
+        .willRespondWith()
+        .status(HttpStatus.OK.value())
+        .body(new PactDslJsonArray().stringType())
+        .toPact();
+  }
+
+  @Test
+  @PactTestFor(pactMethod = "listValidByPolicyInput")
+  public void listingValidRegionsOnAPolicy() throws Exception {
+    var inputs =
+        new TpsPolicyInputs()
+            .inputs(List.of(new TpsPolicyInput().name("test_name").namespace("test_namespace")));
+    var tpsPao =
+        new TpsPaoGetResult()
+            .objectId(existingPolicyId)
+            .attributes(inputs)
+            .effectiveAttributes(inputs);
+    var result = dispatch.listValidRegionsForPao(tpsPao, CloudPlatform.AZURE);
+    assertNotNull(result);
```

Comment:
Most of these are just asserting the result is not null, because we don't generally care about the actual values returned, just that we can deserialize and handle the result (in fact, most of the generators for the responses used will just generate random strings). 
This is largely because a lot of the time we're just passing the value back to the user.

Reply:
Should we care about the shape of the response? Not the values, but the structure?

Reply:
I added a check that we can convert it to the object returned by the API.
But other than that, no - I think that would be testing TPS.